store/storage/pebbledb/db.go

```diff
@@ -0,0 +1,214 @@
+package pebbledb
+
+import (
+	"bytes"
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"math"
+
+	"github.com/cockroachdb/pebble"
+
+	"cosmossdk.io/store/v2"
+)
+
+const (
+	VersionSize = 8
+
+	StorePrefixTpl   = "s/k:%s/"   // s/k:<storeKey>
+	latestVersionKey = "s/_latest" // NB: latestVersionKey key must be lexically smaller than StorePrefixTpl
+)
+
+var (
+	_ store.VersionedDatabase = (*Database)(nil)
+
+	defaultWriteOpts = pebble.Sync
+)
+
+type Database struct {
+	storage *pebble.DB
+}
+
+func New(dataDir string) (*Database, error) {
+	opts := &pebble.Options{
+		Comparer: MVCCComparer,
+	}
+	opts = opts.EnsureDefaults()
+
+	db, err := pebble.Open(dataDir, opts)
+	if err != nil {
+		return nil, fmt.Errorf("failed to open PebbleDB: %w", err)
+	}
+
+	return &Database{
+		storage: db,
+	}, nil
+}
+
+func NewWithDB(storage *pebble.DB) *Database {
+	return &Database{
+		storage: storage,
+	}
+}
+
+func (db *Database) Close() error {
+	return db.storage.Close()
+}
+
+func (db *Database) SetLatestVersion(version uint64) error {
+	var ts [VersionSize]byte
+	binary.LittleEndian.PutUint64(ts[:], version)
+	return db.storage.Set([]byte(latestVersionKey), ts[:], defaultWriteOpts)
+}
+
+func (db *Database) GetLatestVersion() (uint64, error) {
+	bz, closer, err := db.storage.Get([]byte(latestVersionKey))
+	if err != nil {
+		return 0, err
+	}
+
+	if len(bz) == 0 {
+		return 0, closer.Close()
+	}
+
+	return binary.LittleEndian.Uint64(bz), closer.Close()
+}
+
+func (db *Database) Has(storeKey string, version uint64, key []byte) (bool, error) {
+	_, err := getMVCCSlice(db.storage, storeKey, key, version)
+	if err != nil {
+		if errors.Is(err, store.ErrRecordNotFound) {
+			return false, nil
+		}
+
+		return false, fmt.Errorf("failed to perform PebbleDB read: %w", err)
+	}
+
+	return true, nil
+}
+
+func (db *Database) Get(storeKey string, version uint64, key []byte) ([]byte, error) {
+	bz, err := getMVCCSlice(db.storage, storeKey, key, version)
+	if err != nil {
+		if errors.Is(err, store.ErrRecordNotFound) {
+			return nil, nil
+		}
+
+		return nil, fmt.Errorf("failed to perform PebbleDB read: %w", err)
+	}
+
+	return bz, nil
+}
+
+func (db *Database) Set(storeKey string, version uint64, key, value []byte) error {
+	var versionBz [VersionSize]byte
+	binary.LittleEndian.PutUint64(versionBz[:], version)
+
+	batch := db.storage.NewBatch()
+
+	if err := batch.Set([]byte(latestVersionKey), versionBz[:], nil); err != nil {
```

Comment:
I am not sure, every time it is creating a new batch and setting the latest version, is it right?

Reply:
This is correct!

Reply:
OK, so generally we will use the `Batch` of `pebbledb/batch.go`

Reply:
Yes.